File: refactor.js
const fs = require('fs');
const path = require('path');
const Papa = require('papaparse'); // Including papaparse for CSV operations
const JSZip = require('jszip');

function getNewData(data) {
  const producturls = data.map(x => x['url']);

  const new_data = [];
  for (let dt of data) {
    let new_options = dt['options'];
    if (new_options.length > 1) {
      new_options = new_options.slice(1);
    }

    if (dt['url'] !== dt['parent_url'] && dt['parent_url'] !== '') {
      console.log('not matched parent    ', dt['url']);
      try {
        const id = producturls.indexOf(dt['parent_url']);
        new_data.push({
          'options': new_options,
          'brand': dt['brand'],
          'name': data[id]['name'],
          'url': dt['url'],
          'description': dt['description']
        });
      } catch (error) {
        console.log('------- error');
        new_data.push({
          'options': new_options,
          'brand': dt['brand'],
          'name': dt['name'],
          'url': dt['url'],
          'description': dt['description']
        });
      }
    } else {
      new_data.push({
        'options': new_options,
        'brand': dt['brand'],
        'name': dt['name'],
        'url': dt['url'],
        'description': dt['description']
      });
    }
  }

  return new_data;
}

function removeDuplicatedData(data) {
  const skus = {};
  const new_data = [];

  for (let dt of data) {
    skus[dt['name']] = [];
  }

  for (let dt of data) {
    const new_options = dt['options'].filter(op => {
      if (!skus[dt['name']].includes(op['skuNumber'])) {
        skus[dt['name']].push(op['skuNumber']);
        return true;
      }
      return false;
    });

    if (new_options.length > 0) {
      new_data.push({
        'options': new_options,
        'brand': dt['brand'],
        'name': dt['name'],
        'url': dt['url'],
        'description': dt['description']
      });
    }
  }

  return new_data;
}

function refactor(data) {
  let refactoredData = [];
  let count = 0;

  data.forEach(dt => {
    const handle = dt.url.split('https://www.polyperformance.com/')[1];
    const title = dt.name;
    const body = dt.description;
    const vendor = dt.brand;

    dt.options.forEach((option, oid) => {
      const skunumber = option.skuNumber;
      let finalprice = option.finalprice;
      let oldprice = option.oldprice || "0";

      let num_finalprice = parseFloat(finalprice.replace(/,/g, ""));
      let num_oldprice = parseFloat(oldprice.replace(/,/g, ""));
      if (isNaN(num_finalprice)) num_finalprice = 0;

      if (num_oldprice > num_finalprice * 1.7) {
        oldprice = finalprice;
        count++;
      }
      if (num_oldprice === 0 || num_oldprice < num_finalprice) {
        oldprice = finalprice;
      }
      const optionname = option.skuNumber;

      option.images.forEach((img, i) => {
        let tempPd = {
          Handle: handle,
          Title: title,
          "Body (HTML)": body,
          Vendor: vendor,
          "Product Category": "Vehicles & Parts > Vehicle Parts & Accessories",
          "Type": "",
          "Tags": "",
          "Published": "",
          "Option1 Name": "Part #",
          "Option1 Value": optionname,
          "Variant SKU": skunumber,
          "Variant Price": finalprice,
          "Variant Compare At Price": oldprice,
          "Variant Requires Shipping": "TRUE",
          "Variant Taxable": "TRUE",
          "Variant Barcode": "",
          "Image Src": img,
          "Image Position": i + 1,
          "Image Alt Text": "",
          "Gift Card": "",
          "SEO Title": "",
          "SEO Description": "",
          "Google Shopping / Google Product Category": "",
          "Google Shopping / Gender": "",
          "Google Shopping / Age Group": "",
          "Google Shopping / MPN": "",
          "Google Shopping / Condition": "",
          "Google Shopping / Custom Product": "",
          "Google Shopping / Custom Label 0": "",
          "Google Shopping / Custom Label 1": "",
          "Google Shopping / Custom Label 2": "",
          "Google Shopping / Custom Label 3": "",
          "Google Shopping / Custom Label 4": "",
          "Variant Image": "",
          "Variant Weight Unit": "",
          "Variant Tax Code": "",
          "Cost per item": "",
          "Included / United States": "",
          "Price / United States": "",
          "Compare At Price / United States": "",
          "Included / International": "",
          "Price / International": "",
          "Compare At Price / International": "",
          "Status": ""
        };

        // options logic
        if (optionname === "original" || dt.options.length === 1 || oid !== 0 || i !== 0) {
          tempPd['Option1 Name'] = "";
          tempPd['Option1 Value'] = "";
        }

        if (oid !== 0) {
          tempPd.Title = "";
          tempPd["Body (HTML)"] = "";
          tempPd.Vendor = "";
        }

        if (i !== 0) {
          tempPd["Variant SKU"] = "";
          tempPd["Variant Price"] = "";
          tempPd["Variant Compare At Price"] = "";
          tempPd["Product Category"] = ""
          tempPd["Variant Requires Shipping"] = ""
          tempPd["Variant Taxable"] = ""
        }

        refactoredData.push(tempPd);
      });
    });
  });

  console.log(count);
  return refactoredData;
}

function convertToCSV(data, outputPath) {
  // Here you would implement or use a library to write the CSV.
  // Since papaparse is a popular choice, this example will use it.

  const csv = Papa.unparse(data);
  fs.writeFileSync(outputPath, csv, 'utf8');
  console.log(`The JSON data has been successfully converted to '${outputPath}'.`);
}

function zipFile(filePath, outputZipPath, compressionLevel = 'DEFLATE') {
  const zip = new JSZip();
  const fileName = path.basename(filePath);

  fs.readFile(filePath, function (err, data) {
    if (err) throw err;

    // Add the file to the zip
    zip.file(fileName, data, { compression: compressionLevel });

    // Generate the zip file as a buffer
    zip.generateAsync({ type: 'nodebuffer', compression: compressionLevel }).then(function (content) {
      // Write zip file to disk
      fs.writeFile(outputZipPath, content, function (err) {
        if (err) throw err;
        console.log(`Zipped file saved to ${outputZipPath}`);
      });
    });
  });
}

async function getCSV() {
  const brands = JSON.parse(fs.readFileSync("./assets/brands.json", "utf8"));
  let data = [];
  for (let bd of brands) {
    const brandData = JSON.parse(fs.readFileSync(`./assets/data/${bd['name']}.json`, "utf8"));
    data = data.concat(brandData);
  }

  const newData = getNewData(data);
  const removedData = removeDuplicatedData(newData);
  const refactoredData = refactor(removedData);
  convertToCSV(refactoredData, './assets/output.csv');
  zipFile('./assets/output.csv', './assets/output.zip')
}

getCSV().catch(console.error);

File: refactor.py
import os
import json
import pandas as pd
import csv

filenames = os.listdir("./assets/data")

def get_new_data(data):
    producturls = [x['url'] for x in data]

    new_data = []
    for dt in data:
        new_options = dt['options']
        if len(new_options) > 1:
            new_options = new_options[1:]
        
        if dt['url'] != dt['parent_url'] and dt['parent_url'] != '':
            print('not matched parent    ', dt['url'])
            try:
                id = producturls.index(dt['parent_url'])
                new_data.append({
                    'options': new_options,
                    'brand': dt['brand'],
                    'name': data[id]['name'],
                    'url': dt['url'],
                    'description': dt['description']
                })
            except:
                print('------- error')
                new_data.append({
                    'options': new_options,
                    'brand': dt['brand'],
                    'name': dt['name'],
                    'url': dt['url'],
                    'description': dt['description']
                })
        else:
            new_data.append({
                'options': new_options,
                'brand': dt['brand'],
                'name': dt['name'],
                'url': dt['url'],
                'description': dt['description']
            })
    
    return new_data

def remove_duplicated_data(data):
    skus = {}
    for dt in data:
        skus[dt['name']] = []

    new_data = []

    for dt in data:
        new_options = []
        for op in dt['options']:
            if op['skuNumber'] not in skus[dt['name']]:
                new_options.append(op)
                skus[dt['name']].append(op['skuNumber'])
        if (len(new_options) > 0):
            new_data.append({
                'options': new_options,
                'brand': dt['brand'],
                'name': dt['name'],
                'url': dt['url'],
                'description': dt['description']
            })

def refactor(data):
    refactored_data = []
    count = 0
    for dt in data:
        handle = dt["url"].split('https://www.polyperformance.com/')[1]
        title = dt["name"]
        body = dt["description"]
        vendor = dt["brand"]

        for oid, option in enumerate(dt["options"]):
            skunumber = option["skuNumber"]
            finalprice = option["finalprice"]
            oldprice = option["oldprice"]

            if oldprice == "":
                oldprice = "0"
            try:
                num_finalprice = float(finalprice.replace(",", ""))
            except:
                num_finalprice = 0
                pass
            num_oldprice = float(oldprice.replace(",", ""))

            if num_oldprice > num_finalprice * 1.7:
                oldprice = finalprice
                count += 1
            if num_oldprice == 0 or num_oldprice < num_finalprice:
                oldprice = finalprice
            optionname = option["skuNumber"]


            for i, img in enumerate(option["images"]):
                temp_pd = {
                    "Handle": handle,
                    "Title": title,
                    "Body (HTML)": body,
                    "Vendor": vendor,
                    "Product Category": "",
                    "Type": "",
                    "Tags": "",
                    "Published": "",
                    "Option1 Name": "Part #",
                    "Option1 Value": optionname,
                    "Variant SKU": skunumber,
                    "Variant Price": finalprice,
                    "Variant Compare At Price": oldprice,
                    "Variant Requires Shipping": "",
                    "Variant Taxable": "",
                    "Variant Barcode": "",
                    "Image Src": img,
                    "Image Position": i+1,
                    "Image Alt Text": "",
                    "Gift Card": "",
                    "SEO Title": "",
                    "SEO Description": "",
                    "Google Shopping / Google Product Category": "",
                    "Google Shopping / Gender": "",
                    "Google Shopping / Age Group": "",
                    "Google Shopping / MPN": "",
                    "Google Shopping / Condition": "",
                    "Google Shopping / Custom Product": "",
                    "Google Shopping / Custom Label 0": "",
                    "Google Shopping / Custom Label 1": "",
                    "Google Shopping / Custom Label 2": "",
                    "Google Shopping / Custom Label 3": "",
                    "Google Shopping / Custom Label 4": "",
                    "Variant Image": "",
                    "Variant Weight Unit": "",
                    "Variant Tax Code": "",
                    "Cost per item": "",
                    "Included / United States": "",
                    "Price / United States": "",
                    "Compare At Price / United States": "",
                    "Included / International": "",
                    "Price / International": "",
                    "Compare At Price / International": "",
                    "Status": ""
                }

                if optionname == "original":
                    temp_pd['Option1 Name'] = ""
                    temp_pd['Option1 Value'] = ""

                if len(dt["options"]) == 1:
                    temp_pd['Option1 Name'] = ""
                    temp_pd['Option1 Value'] = ""


                if oid == 0 and i == 0:
                    refactored_data.append(temp_pd)
                else:
                    temp_pd['Option1 Name'] = ""
                    temp_pd["Title"] = ""
                    temp_pd["Body (HTML)"] = ""
                    temp_pd["Vendor"] = ""

                    if i != 0:
                        temp_pd['Option1 Name'] = ""
                        temp_pd['Option1 Value'] = ""
                        temp_pd["Variant SKU"] = ""
                        temp_pd["Variant Price"] = ""
                        temp_pd["Variant Compare At Price"] = ""

                    refactored_data.append(temp_pd)
    print(count)
    return refactored_data


def convert(data, output_file_path):
    # Check if data is a list of records (dictionaries)
    if isinstance(data, list) and all(isinstance(item, dict) for item in data):
        # Get the headers (keys) for the CSV file from the first JSON object
        headers = list(data[0].keys())

        # Write to the CSV file
        with open(output_file_path, 'w', newline='', encoding='utf-8') as csv_file:
            writer = csv.DictWriter(csv_file, fieldnames=headers)

            # Write the header row
            writer.writeheader()
            
            # Write each data row
            for row in data:
                writer.writerow(row)
    else:
        print("JSON data is not a list of dictionaries.")

    print(f"The JSON data has been successfully converted to '{output_file_path}'.")

def get_csv():
    data = []
    for fn in filenames:
        with open(f"./assets/data/{fn}", "r", encoding="utf-8") as f:
            data += json.load(f)

    new_data = get_new_data(data)
    removed_data = remove_duplicated_data(data)
    refactored_data = refactor(new_data)
    convert(refactored_data, './assets/output.csv')

get_csv()